doc/api/stream.md

```diff
@@ -1112,22 +1290,112 @@ class SimpleProtocol extends Readable {
+
+*Note*: The `stream.Duplex` class prototypically inherits from `stream.Readable`
+and parasitically from `stream.Writable`.
+
+Custom Duplex streams *must* call the `new stream.Duplex([options])`
+constructor and implement *both* the `readable.\_read()` and
+`writable.\_write()` methods.
+
+#### new stream.Duplex(options)
+
+* `options` {Object} Passed to both Writable and Readable
+  constructors. Also has the following fields:
+  * `allowHalfOpen` {Boolean} Default = `true`. If set to `false`, then
+    the stream will automatically end the readable side when the
+    writable side ends and vice versa.
+  * `readableObjectMode` {Boolean} Default = `false`. Sets `objectMode`
+    for readable side of the stream. Has no effect if `objectMode`
+    is `true`.
+  * `writableObjectMode` {Boolean} Default = `false`. Sets `objectMode`
+    for writable side of the stream. Has no effect if `objectMode`
+    is `true`.
+
+For example:
+
+```js
+const Duplex = require('stream').Duplex;
+
+class MyDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+  }
+}
+```
+
+Or, when using pre-ES6 style constructors:
+
+```js
+const Duplex = require('stream').Duplex;
+const util = require('util');
+
+function MyDuplex(options) {
+  if (!(this instanceof MyDuplex))
+    return new MyDuplex(options);
+  Duplex.call(this, options);
+}
+util.inherits(MyDuplex, Duplex);
+```
+
+#### An Example Duplex Stream
+
+The following illustrates a simple example of a Duplex stream that buffers
+incoming written data via the [Writable][] interface that is read back out
+via the [Readable][] interface.
+
+```js
+const Duplex = require('stream').Duplex;
+const kBuf = Symbol('buffer');
+
+class MyDuplex extends Duplex {
+  constructor(options) {
+    super(options);
+  }
+
+  _write(chunk, encoding, next) {
+    var buffer = this[kBuf];
+    if (!Array.isArray(buffer))
+      buffer = this[kBuf] = [];
+    buffer.push(chunk);
+    next();
+  }
+
+  _read(size) {
+    var buffer = this[kBuf];
+    if (Array.isArray(buffer))
+      buffer.forEach((chunk) => this.push(chunk));
+  }
+}
+```
+
+### Implementing a Transform Stream
 
-A "transform" stream is a duplex stream where the output is causally
-connected in some way to the input, such as a [zlib][] stream or a
-[crypto][] stream.
+A [Transform][] stream is a [Duplex][] stream where the output is computed
+in some way from the input. Examples include [zlib][] streams or [crypto][]
+streams that compress, encrypt, or decrypt data.
 
-There is no requirement that the output be the same size as the input,
+*Note*: There is no requirement that the output be the same size as the input,
 the same number of chunks, or arrive at the same time. For example, a
 Hash stream will only ever have a single chunk of output which is
-provided when the input is ended. A zlib stream will produce output
+provided when the input is ended. A `zlib` stream will produce output
 that is either much smaller or much larger than its input.
```

Comment:
We should note that if you're using a transform stream, your application should take care to provide a mechanism for taking ownership of the data that it generates — if you pipe to it endlessly without reading from it, correctly implemented transform streams will pause the writable side due to backpressure from the readable side.